### PR TITLE
fix(stepper): Selection methods should ignore disabled items

### PR DESCRIPTION
**Pull request by driskull**
**Related Issue:** #4315 

## Summary

fix(stepper): Selection methods should ignore disabled items. (#4315)

Updates prevStep, nextStep, startStep, and endStep methods to filter out selection of disabled stepper items.

### CODE REVIEW

**Comment by driskull** [comment]
@jcfranco I opened a separate PR with just the refactoring. Once that merges this one will be a bit simpler.

https://github.com/Esri/calcite-components/pull/4491

**Review by jcfranco — APPROVED** [review]
LGTM! 💥

Aside from the test formatting helper comment, the rest could be done in a follow-up PR.

File: src/components/stepper/stepper.e2e.ts [hunk 1]
```diff
@@ -165,6 +165,39 @@ describe("calcite-stepper", () => {
     expect(step4).not.toHaveAttribute("active");
   });
 
+  it("navigates disabled items correctly with nextStep and prevStep methods", async () => {
+    const page = await newE2EPage();
+    await page.setContent(`
+    <calcite-stepper>
```
[the remark this answers is not in the document]
Comment (driskull) [reply src/components/stepper/stepper.e2e.ts:171]: I can do this in followup

File: src/components/stepper/stepper.tsx [hunk 2]
```diff
@@ -219,10 +241,14 @@ export class Stepper {
   //
   //--------------------------------------------------------------------------
 
-  /** created list of Stepper items */
-  private items = [];
+  /** map of Stepper items */
```
[the remark this answers is not in the document]
Comment (driskull) [reply src/components/stepper/stepper.tsx:244]: I can do this in followup

File: src/components/stepper/stepper.tsx [hunk 3]
```diff
@@ -177,39 +175,63 @@ export class Stepper {
...
-      this.currentPosition + 1 < this.items.length
-        ? this.currentPosition + 1
-        : this.currentPosition;
+    const enabledStepIndex = this.getEnabledStepIndex(this.currentPosition + 1, "next");
+
+    if (typeof enabledStepIndex !== "number") {
+      return;
+    }
+
+    this.currentPosition = enabledStepIndex;
```
[the remark this answers is not in the document]
Comment (driskull) [reply src/components/stepper/stepper.tsx:184]: I can do this in followup

**Comment by benelan** [comment]
Possibly a couple unstable stepper tests. Or it could be an issue with the container.
```bash
FAIL src/components/stepper/stepper.e2e.ts (75.272 s)
  ● calcite-stepper › navigates disabled items correctly with nextStep and prevStep methods

    : Timeout - Async callback was not invoked within the 30000 ms timeout specified by jest.setTimeout.Timeout - Async callback was not invoked within the 30000 ms timeout specified by jest.setTimeout.Error:

      166 |   });
      167 |
    > 168 |   it("navigates disabled items correctly with nextStep and prevStep methods", async () => {
          |   ^
      169 |     const page = await newE2EPage();
      170 |     await page.setContent(`
      171 |     <calcite-stepper>

      at new Spec (node_modules/jest-jasmine2/build/jasmine/Spec.js:116:22)
      at Suite.<anonymous> (src/components/stepper/stepper.e2e.ts:168:3)

  ● calcite-stepper › navigates disabled items correctly with nextStep and prevStep methods

    Timeout - Async callback was not invoked within the 30000 ms timeout specified by jest.setTimeout.Error: Timeout - Async callback was not invoked within the 30000 ms timeout specified by jest.setTimeout.

      at mapper (node_modules/jest-jasmine2/build/queueRunner.js:27:45)

  ● calcite-stepper › navigates correctly with startStep and endStep methods

    Protocol error (Runtime.callFunctionOn): Target closed.

      187 |     expect(step2).not.toHaveAttribute("active");
      188 |     expect(step3).not.toHaveAttribute("active");
    > 189 |     await element.callMethod("nextStep");
          |                   ^
      190 |     await page.waitForChanges();
      191 |     expect(step1).not.toHaveAttribute("active");
      192 |     expect(step2).not.toHaveAttribute("active");

      at node_modules/puppeteer/src/common/Connection.ts:291:57
      at CDPSession.send (node_modules/puppeteer/src/common/Connection.ts:290:12)
      at ExecutionContext._evaluateInternal (node_modules/puppeteer/src/common/ExecutionContext.ts:254:44)
      at ExecutionContext.evaluate (node_modules/puppeteer/src/common/ExecutionContext.ts:140:23)
      at E2EElement.e2eRunActions (node_modules/@stencil/core/testing/index.js:3544:61)
      at E2EElement.callMethod (node_modules/@stencil/core/testing/index.js:3311:12)
      at Object.<anonymous> (src/components/stepper/stepper.e2e.ts:189:19)
          at runMicrotasks (<anonymous>)

  ● calcite-stepper › navigates disabled items correctly with startStep and endStep methods

    expected to not have the attribute "active"

      268 |     await page.waitForChanges();
      269 |     expect(step1).not.toHaveAttribute("active");
    > 270 |     expect(step2).not.toHaveAttribute("active");
          |                       ^
      271 |     expect(step3).toHaveAttribute("active");
      272 |     expect(step4).not.toHaveAttribute("active");
      273 |     await element.callMethod("startStep");

      at Object.<anonymous> (src/components/stepper/stepper.e2e.ts:270:23)
          at runMicrotasks (<anonymous>)
```

File: src/components/stepper/stepper.tsx [hunk 4]
```diff
@@ -236,6 +262,23 @@ export class Stepper {
...
 
+  private getEnabledStepIndex(
+    startIndex: number,
+    direction: "next" | "previous" = "next"
+  ): number | null {
+    const { items, currentPosition } = this;
+
+    let newIndex = startIndex;
+
+    while (items[newIndex]?.disabled) {
```
[the remark this answers is not in the document]
Comment (driskull) [reply src/components/stepper/stepper.tsx:273]: I can do this in followup

[the remark this answers is not in the document]
Comment (driskull) [reply src/components/stepper/stepper.tsx:273]: To do this, would we also need to restrict `goToStep()` to only allow enabled steps? Which direction would it go if the step was disabled? Would it throw an error?

[the remark this answers is not in the document]
Comment (driskull) [reply src/components/stepper/stepper.tsx:273]: @jcfranco can you create a separate issue if we want to do this?

Currently, the event emits the position index from all items including disabled ones.

**Pull request merged**